runtime/compiler/z/runtime/Emulation.cpp

```diff
@@ -0,0 +1,251 @@
+   dx2 = dxl2 | (dxh2 << 12);
+   shift = (start[5] & 0xE) >> 1;
+   isLogical = (start[5] & 1) == 1;
+   }
+
+void LXAEmulator::emulate(mcontext_t *cpu)
+   {
+   int64_t addr;
+   int32_t tmp = (int32_t) dx2;
+
+   // sign extend immediate
+   if (tmp & 0x80000)
+      {
+      tmp |= 0xFFF00000;
+      }
+
+   if (x2 != 0)
+      {
+      tmp += cpu->gregs[x2];
+      }
+
+   if (isLogical)
+      {
+      addr = (int64_t)(uint32_t) tmp;
+      }
+   else
+      {
+      addr = (int64_t) tmp;
+      }
+
+   addr <<= shift;
+   addr += cpu->gregs[b2];
+   cpu->gregs[r1] = addr;
+   }
+
+class BDEPGEmulator : public InstEmulator
+   {
+   private:
+      uint8_t r1, r2, r3;
+   public:
+      BDEPGEmulator(uint8_t *start);
+      virtual void emulate(mcontext_t *cpu);
+   };
+
+BDEPGEmulator::BDEPGEmulator(uint8_t *start)
+   {
+   r1 = (start[3]&0xF0) >> 4;
+   r2 = start[3]&0x0F;
+   r3 = (start[2]&0xF0) >> 4;
+   }
+
+void BDEPGEmulator::emulate(mcontext_t *cpu)
+   {
+   uint64_t val = cpu->gregs[r2];
+   uint64_t mask = cpu->gregs[r3];
+   uint64_t res = 0;
+
+   for (int n = 0; mask; n++)
+      {
+      if (mask & (1ULL << 63))
+         {
+         res |= (val & (1ULL << 63)) >> n;
+         val <<= 1;
+         }
+      mask <<= 1;
+      }
+
+   cpu->gregs[r1] = res;
+   }
+
+class BEXTGEmulator : public InstEmulator
+   {
+   private:
+      uint8_t r1, r2, r3;
+   public:
+      BEXTGEmulator(uint8_t *start);
+      virtual void emulate(mcontext_t *cpu);
+   };
+
+BEXTGEmulator::BEXTGEmulator(uint8_t *start)
+   {
+   r1 = (start[3]&0xF0) >> 4;
+   r2 = start[3]&0x0F;
+   r3 = (start[2]&0xF0) >> 4;
+   }
+
+void BEXTGEmulator::emulate(mcontext_t *cpu)
+   {
+   uint64_t val = cpu->gregs[r2];
+   uint64_t mask = cpu->gregs[r3];
+   uint64_t res = 0;
+
+   for (int k = 0; mask; mask <<= 1, val <<= 1)
+      {
+      if (mask & (1ULL<<63))
+         {
+         res |= (val & (1ULL<<63)) >> k;
+         k++;
+         }
+      }
+
+   cpu->gregs[r1] = res;
+   }
+
+class CLZGEmulator : public InstEmulator
+   {
+   private:
+      uint8_t r1, r2;
+   public:
+      CLZGEmulator(uint8_t *start);
+      virtual void emulate(mcontext_t *cpu);
+   };
+
+CLZGEmulator::CLZGEmulator(uint8_t *start)
+   {
+   r1 = (start[3]&0xF0) >> 4;
+   r2 = start[3]&0x0F;
+   }
+
+void CLZGEmulator::emulate(mcontext_t *cpu)
+   {
+   cpu->gregs[r1] = (uint64_t)leadingZeroes(cpu->gregs[r2]);
+   }
+
+class CTZGEmulator : public InstEmulator
+   {
+   private:
+      uint8_t r1, r2;
+   public:
+      CTZGEmulator(uint8_t *start);
+      virtual void emulate(mcontext_t *cpu);
+   };
+
+CTZGEmulator::CTZGEmulator(uint8_t *start)
+   {
+   r1 = (start[3]&0xF0) >> 4;
+   r2 = start[3]&0x0F;
+   }
+
+void CTZGEmulator::emulate(mcontext_t *cpu)
+   {
+   cpu->gregs[r1] = (uint64_t)trailingZeroes(cpu->gregs[r2]);
+   }
+
+InstEmulator *InstEmulator::decode(uint8_t *pc)
+   {
+   if (pc[-6] == 0xE3 && (pc[-1]&0xF0) == 0x60 && (pc[-1]&0x0F) < 10)
```

Comment:
I think this deserves a comment to explain the logic. Specifically, we are matching all the LXA* instructions with opcodes from 0xE360 to 0xE369.  Please also add spaces before and after the `&`, and add parens around each of the `&&` operands.

`if ((pc[-6] == 0xE3) && ((pc[-1] & 0xF0) == 0x60) && ((pc[-1] & 0x0F) < 10))`

Reply:
Fixed in https://github.com/eclipse-openj9/openj9/pull/20507/commits/7635eb9a3e481ca198c22e8a0c260fc13c48d310